src/gui/TextEditor.cpp

```diff
@@ -904,8 +905,66 @@ auto TextEditor::blinkCallback(TextEditor* te) -> gint {
     return false;
 }
 
+void TextEditor::setTextToPangoLayout(PangoLayout* pl) const {
+    if (!this->preeditString.empty()) {
+        GtkTextIter cursorIter = {nullptr};
+        GtkTextMark* cursor = gtk_text_buffer_get_insert(this->buffer);
+        gtk_text_buffer_get_iter_at_mark(this->buffer, &cursorIter, cursor);
+
+        string text = this->text->getText();
+        int offset = gtk_text_iter_get_offset(&cursorIter);
+        int pos = gtk_text_iter_get_line_index(&cursorIter);
+
+        for (gtk_text_iter_set_line_index(&cursorIter, 0); gtk_text_iter_backward_line(&cursorIter);) {
+            pos += gtk_text_iter_get_bytes_in_line(&cursorIter);
+        }
+        gtk_text_iter_set_offset(&cursorIter, offset);
+        string txt = text.substr(0, pos) + preeditString + text.substr(pos);
+
+        PangoAttrList* attrlist = pango_attr_list_new();
+        PangoAttrList* preedit_attrlist = this->preeditAttrList;
+        pango_attr_list_splice(attrlist, preedit_attrlist, pos, preeditString.length());
+        pango_layout_set_attributes(pl, attrlist);
+        pango_attr_list_unref(attrlist);
+        attrlist = nullptr;
+        pango_layout_set_text(pl, txt.c_str(), txt.length());
+    } else {
+        string txt = this->text->getText();
+        pango_layout_set_text(pl, txt.c_str(), txt.length());
+    }
+}
+
+auto TextEditor::computeBoundingRect() -> Rectangle<double> {
+    /**
```

Comment:
nitpick: double `*` should be reserved for documentation comments only.
```suggestion
    /*
```